src/tls13.rs

```diff
@@ -0,0 +1,313 @@
+        assert_eq!(key.as_ref().len(), CHACHA_KEY_LENGTH); // ChaCha key length must be 32 bytes.
+        let mut chacha_key = [0u8; CHACHA_KEY_LENGTH];
+        chacha_key[..CHACHA_KEY_LENGTH].copy_from_slice(key.as_ref());
+
+        Box::new(Tls13ChaCha20Poly1305 {
+            key: chacha_key,
+            iv: iv,
+        })
+    }
+
+    fn decrypter(&self, key: AeadKey, iv: Iv) -> Box<dyn MessageDecrypter> {
+        assert_eq!(key.as_ref().len(), CHACHA_KEY_LENGTH); // ChaCha key length must be 32 bytes.
+        let mut chacha_key = [0u8; CHACHA_KEY_LENGTH];
+        chacha_key[..CHACHA_KEY_LENGTH].copy_from_slice(key.as_ref());
+
+        Box::new(Tls13ChaCha20Poly1305 {
+            key: chacha_key,
+            iv: iv,
+        })
+    }
+
+    fn key_len(&self) -> usize {
+        CHACHA_KEY_LENGTH // ChaCha key must be 32 bytes.
+    }
+
+    fn extract_keys(
+        &self,
+        key: AeadKey,
+        iv: Iv,
+    ) -> Result<ConnectionTrafficSecrets, UnsupportedOperationError> {
+        Ok(ConnectionTrafficSecrets::Chacha20Poly1305 { key, iv })
+    }
+}
+
+/// [`MessageEncrypter`] for ChaCha 1.3
+/// the [`payload`] field that comes from the [`BorrowedPlainMessage`] is structured to include the message which is an arbitrary length,
+/// an encoding type that is 1 byte and then finally the tag which is 16 bytes.
+/// ex : [1, 2, 3, 5, 6, 7, 8, 9, 10, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12 ,13, 14, 15, 16]
+///       ^                       ^   ^  ^                                                   ^
+///            Message (N bytes)   Encoding (1 Byte)              Tag (16 bytes)
+impl MessageEncrypter for Tls13ChaCha20Poly1305 {
+    fn encrypt(&self, msg: BorrowedPlainMessage, seq: u64) -> Result<OpaqueMessage, rustls::Error> {
+        // Adding the size of message, the tag and encoding type to the capacity of payload vector.
+        let total_len = msg.payload.len() + 1 + CHACHA_TAG_LENGTH;
+        let mut payload = Vec::with_capacity(total_len);
+
+        // Construct payload.
+        payload.extend_from_slice(msg.payload);
+        payload.push(msg.typ.get_u8());
+
+        // Set up needed parameters for ChaCha encrypt.
+        let nonce = Nonce::new(&self.iv, seq);
+        let auth_data = make_tls13_aad(total_len);
+        let mut tag = [0u8; CHACHA_TAG_LENGTH];
+
+        // Encrypting the payload in place. +1 is added to account for encoding type that must also be encrypted.
+        match chacha20_poly1305_encrypt_in_place(
+            &self.key,
+            &nonce.0,
+            &auth_data,
+            &mut payload[..msg.payload.len() + 1],
+            &mut tag,
+        ) {
+            Ok(_) => {
+                payload.extend_from_slice(&tag); // Add tag to the end of the payload.
+                Ok(OpaqueMessage::new(
+                    rustls::ContentType::ApplicationData,
+                    rustls::ProtocolVersion::TLSv1_3,
```

Comment:
>rustls::ProtocolVersion::TLSv1_3, [](http://example.com/codeflow?start=20&length=33)

For ring I'm seeing:
ProtocolVersion::TLSv1_2,

Which doesn't seem correct??

Reply:
Yes i noticed this as well, Should we change to TLSv1_2 ? or keep as TLSv1_3